proof/refine/X64/Invariants_H.thy

```diff
@@ -1128,17 +1128,29 @@ definition
 where
   "valid_x64_irq_state' irqState \<equiv> \<forall>irq > maxIRQ. irqState irq = X64IRQFree"
 
-definition
-  valid_arch_state' :: "kernel_state \<Rightarrow> bool"
-where
+definition valid_ioapic_2 :: "machine_word \<Rightarrow> (machine_word \<Rightarrow> 8 word) \<Rightarrow> bool" where
+  "valid_ioapic_2 num_ioapics ioapic_nirqs \<equiv>
+    num_ioapics \<le> of_nat Kernel_Config.maxNumIOAPIC \<and>
```

Comment:
Sanity check: if you unfold `Kernel_Config.maxNumIOAPIC` in CRefine without a localised unfold in a lemma(s) that restrict the unfold to only generic properties we need, that is a pattern that prevents future parametrisation of that `Kernel_Config` entry. It's survivable for X64, but we should try avoid this pattern in principle (though this still happens in AARCH64, so it's clearly not completely escapable). Thoughts?

Reply:
I was going to go the lazy way, because X64, but you're right that we should get into the habit. Will have a look.

Reply:
Ok, I managed to reduce the unfoldings by one. The other two are for array guards where `maxNumIOAPIC` corresponds to the dimension of the array, so the unfolding should be guaranteed to succeed. I've isolated them a little bit more.